File: code/src/Yolov4.cpp
#include "Yolov4.h"

YOLOv4::YOLOv4(const YAML::Node &config) : Detection(config) {
    anchors = config["anchors"].as<std::vector<std::vector<int>>>();
    int index = 0;
    for (const int &stride : strides)
    {
        grids.push_back({num_anchors[index], int(IMAGE_HEIGHT / stride), int(IMAGE_WIDTH / stride)});
    }
    refer_rows = 0;
    refer_cols = 6;
    for (const std::vector<int> &grid : grids) {
        refer_rows += std::accumulate(grid.begin(), grid.end(), 1, std::multiplies<int>());
    }
    GenerateReferMatrix();
    letter_box = resize == "keep_ratio";
    class_colors.resize(CATEGORY);
    srand((int) time(nullptr));
    for (cv::Scalar &class_color : class_colors)
        class_color = cv::Scalar(rand() % 255, rand() % 255, rand() % 255);
}

void YOLOv4::GenerateReferMatrix() {
    refer_matrix = cv::Mat(refer_rows, refer_cols, CV_32FC1);
    int position = 0;
    for (int n = 0; n < (int)grids.size(); n++) {
        for (int c = 0; c < grids[n][0]; c++) {
            std::vector<int> anchor = anchors[n * grids[n][0] + c];
            for (int h = 0; h < grids[n][1]; h++)
                for (int w = 0; w < grids[n][2]; w++) {
                    auto *row = refer_matrix.ptr<float>(position);
                    row[0] = w;
                    row[1] = grids[n][2];
                    row[2] = h;
                    row[3] = grids[n][1];
                    row[4] = anchor[0];
                    row[5] = anchor[1];
                    position++;
                }
        }
    }
}

std::vector<DetectRes> YOLOv4::PostProcess(const std::vector<cv::Mat> &vec_Mat, float *output) {
    std::vector<DetectRes> vec_result;
    int index = 0;
    for (const cv::Mat &src_img : vec_Mat) {
        DetectRes result;
        float *out = output + index * outSize;
        float ratio = std::max(float(src_img.cols) / float(IMAGE_WIDTH), float(src_img.rows) / float(IMAGE_HEIGHT));
        cv::Mat result_matrix = cv::Mat(refer_rows, CATEGORY + 5, CV_32FC1, out);
        for (int row_num = 0; row_num < refer_rows; row_num++) {
            Bbox box{};
            auto *row = result_matrix.ptr<float>(row_num);
            if ((letter_box ? row[4] : sigmoid(row[4])) < obj_threshold)
                continue;
            auto max_pos = std::max_element(row + 5, row + CATEGORY + 5);
            box.prob = letter_box ? row[4] * row[max_pos - row] : sigmoid(row[4]) * sigmoid(row[max_pos - row]);
            box.classes = max_pos - row - 5;
            auto *anchor = refer_matrix.ptr<float>(row_num);
            box.x = letter_box ? (float)(row[0] * 2 - 0.5 + anchor[0]) / anchor[1] * (float)IMAGE_WIDTH * ratio : (sigmoid(row[0]) + anchor[0]) / anchor[1] * (float)src_img.cols;
            box.y = letter_box ? (float)(row[1] * 2 - 0.5 + anchor[2]) / anchor[3] * (float)IMAGE_HEIGHT * ratio : (sigmoid(row[1]) + anchor[2]) / anchor[3] * (float)src_img.rows;
            box.w = letter_box ? (float)pow(row[2] * 2, 2) * anchor[4] * ratio : exp(row[2]) * anchor[4] / (float)IMAGE_WIDTH * (float)src_img.cols;
            box.h = letter_box ? (float)pow(row[3] * 2, 2) * anchor[5] * ratio : exp(row[3]) * anchor[5] / (float)IMAGE_HEIGHT * (float)src_img.rows;
            result.det_results.push_back(box);
        }
        NmsDetect(result.det_results);
        vec_result.push_back(result);
        index++;
    }
    return vec_result;
}

float YOLOv4::sigmoid(float in){
    return 1.f / (1.f + exp(-in));
}

File: python/models/YOLOv4.py
import numpy as np
import torch
from torchvision.ops import nms
from .Detection import Detection, DetectRes, Bbox


class Yolov4(Detection):
    def __init__(self, config):
        super().__init__(config)
        self.anchors = config['anchors']
        self.grids, self.refer_rows, self.refer_cols = [], 0, 6
        for index, stride in enumerate(self.strides):
            self.grids.append([self.num_anchors[index], self.IMAGE_HEIGHT // stride, self.IMAGE_WIDTH // stride])
            self.refer_rows += self.num_anchors[index] * self.IMAGE_HEIGHT // stride * self.IMAGE_WIDTH // stride
        self.refer_matrix = np.zeros((self.refer_rows, self.refer_cols))
        self.generate_refer_matrix()
        self.letter_box = self.resize == "keep_ratio"
        self.output_shape = (self.BATCH_SIZE, self.num_rows, self.CATEGORY + 5)

    def generate_refer_matrix(self):
        position = 0
        for n in range(len(self.grids)):
            for c in range(self.grids[n][0]):
                anchor = self.anchors[n * self.grids[n][0] + c]
                for h in range(self.grids[n][1]):
                    for w in range(self.grids[n][2]):
                        row = self.refer_matrix[position]
                        row[0] = w
                        row[1] = self.grids[n][2]
                        row[2] = h
                        row[3] = self.grids[n][1]
                        row[4] = anchor[0]
                        row[5] = anchor[1]
                        position += 1

    def parse_boxes(self, x, anchors):
        # Convert nx4 boxes from [x, y, w, h] to [x1, y1, x2, y2] where xy1=top-left, xy2=bottom-right
        y = torch.from_numpy(x.copy())
        y[:, 0] = (y[:, 0] * 2 - 0.5 + anchors[:, 0]) / anchors[:, 1] * self.IMAGE_WIDTH if self.letter_box else \
            (torch.sigmoid(y[:, 0]) + anchors[:, 0]) / anchors[:, 1]
        y[:, 1] = (y[:, 1] * 2 - 0.5 + anchors[:, 2]) / anchors[:, 3] * self.IMAGE_HEIGHT if self.letter_box else \
            (torch.sigmoid(y[:, 1]) + anchors[:, 2]) / anchors[:, 3]
        y[:, 2] = torch.pow(y[:, 2] * 2, 2) * anchors[:, 4] if self.letter_box else \
            (torch.exp(y[:, 2]) * anchors[:, 4]) / self.IMAGE_WIDTH
        y[:, 3] = torch.pow(y[:, 3] * 2, 2) * anchors[:, 5] if self.letter_box else \
            (torch.exp(y[:, 3]) * anchors[:, 5]) / self.IMAGE_HEIGHT
        return y

    def post_process(self, image_batch, outputs):
        vec_result = []
        max_wh = 7680
        outputs = outputs[0].reshape(self.output_shape)
        for src_img, output in zip(image_batch, outputs):
            result = DetectRes()
            height, width = src_img.shape[:2]
            ratio = max(width / self.IMAGE_WIDTH, height / self.IMAGE_HEIGHT)
            if not self.letter_box:
                output[:, 4:] = torch.sigmoid(torch.from_numpy(output[:, 4:])).numpy()
            index = output[:, 4] > self.obj_threshold
            conf, j = torch.from_numpy(output[index, 5:]).max(dim=1)
            conf *= output[index, 4]
            boxes = self.parse_boxes(output[index, :4], torch.from_numpy(self.refer_matrix[index]))
            if self.letter_box:
                boxes = boxes * ratio
            else:
                boxes[:, 0::2], boxes[:, 1::2] = boxes[:, 0::2] * width, boxes[:, 1::2] * height
            boxes = self.xywh2xyxy(boxes)
            c = j * (0 if self.agnostic else max_wh)  # classes
            output_indexes = nms(boxes + c.unsqueeze(1), conf, self.nms_threshold)  # NMS
            for output_index in output_indexes:
                bbox = Bbox(j[output_index], conf[output_index], boxes[output_index, 0], boxes[output_index, 1],
                            boxes[output_index, 2] - boxes[output_index, 0],
                            boxes[output_index, 3] - boxes[output_index, 1])
                result.det_results.append(bbox)
            vec_result.append(result)
        return vec_result
